Add actions for draw and forfeit

kellylougheed opened the pull request:
IN PROGRESS! So far...
- Added buttons to for players to draw, reject a draw, and forfeit on the show page.
- Added user authentication before most actions (except the index page) and updated tests accordingly.

Still need to...
- Add web sockets.
- Write TESTS!!!!

Question: I know how to make JavaScript show messages on the screen, but how do I make the JavaScript show messages to specific users (one player but not the other)?

lasley commented:
The JavaScript will only ever be acting in the context of the user, because it is processed in the browser. 

You won't be able to actually push the message to the user, because that's going to be websocket. You can get it partially ready though by:
* create route for message
* JS function to hit route with AJAX
* jQuery `$(document).ready(` to call the JS function and display message
* Set timeout to call js function at intervals (poll)

kellylougheed commented:
Thanks for answering the question @lasley! By routes, do you mean setting up a method in the controller for messages specifically? I made methods for draw, reject_draw, and forfeit.

lasley commented:
Yup exactly. Otherwise, you would have to use a pre-existing route and it would expand the scope of that method for no reason. Better to just have a /game/1/message or whatever

kellylougheed commented:
OK, I added a message route and some pretty skeletal JavaScript! I'm pretty unclear on how the JS + the Rails database are going to interact to show a message/buttons to a specific user under certain conditions. Will that be cleared up with the sockets?

lasley commented on app/assets/javascripts/draw-forfeit.js:
```diff
@@ -0,0 +1,58 @@
+$(function() {
+  $(".request-draw").click(function() {
+    var updateURL = $(event.target).data("update-url");
```
Probably wannt `event.currentTarget` https://developer.mozilla.org/en-US/docs/Web/API/Event/Comparison_of_Event_Targets

lasley commented on app/assets/javascripts/draw-forfeit.js:
```diff
@@ -0,0 +1,58 @@
+  });
+
+  $(".accept-draw").click(function() {
+    var updateURL = $(event.target).data("update-url");
```
Where is `event` defined?

lasley commented on app/controllers/games_controller.rb:
```diff
@@ -73,12 +75,70 @@ def add_player
+      @game.update("#{@color}_draw" => true)
+      flash[:info] = "Player #{current_player.id} has asked for a draw. Player #{@opponent_id}, you may accept or reject the draw."
+    elsif !@game.white.draw
+      @game.update(:white_draw => true)
```
You shouldn't mutate records directly in controllers

kellylougheed replied:
OK, making some methods in the game model!

lasley commented:
Yeah this looks good for the structure. In terms of the JS, we will basically just replace the AJAX that exists with a push to the web socket instead. They operate under similar principles in terms of data transfer, just a different delivery method (UPS instead of USPS)

kellylougheed commented:
I would love some thoughts and opinions on this PR. I'm integrating Pusher (which has worked, except for a timestamp error I got later in the process) but I'm not sure quite the best way to go. I feel like there are two paths...

- Dynamically update the HTML (as seen in application.js.coffee) in terms of showing messages and buttons. The event would have to be triggered in the controller depending on the player's perspective (if their opponent requested a draw, or if they did).

- Somehow use a "message" model and controller. I already created this, but now I'm second-guessing it because it seems like it could get messy compared to dynamically generated HTML. Presumably the controller would generate messages that would trigger an event after their creation and show up on the appropriate user's screen.

I'm not positive how to implement either of these and would love suggestions and ideas.

lasley commented:
Definitely dynamic, although typically you have a controller too. You want to use a JSON response in the controller, ingest the data via the AJAX & act upon it.

kellylougheed commented:
So have the original AJAX call render a success message that the controller can read and act on? I think that makes sense. Thanks for the pointer!

Sent from my iPhone

> On Dec 17, 2016, at 12:44 PM, Dave Lasley <notifications@github.com> wrote:
> 
> Definitely dynamic, although typically you have a controller too. You want to use a JSON response in the controller, ingest the data via the AJAX & act upon it.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

kellylougheed commented:
@lasley Just saw your comment addendum & reset my branch back to before I went on a controller-creating rampage. Trying to envision how a message route in "Game" would fit in -- would the "success" function for my initial AJAX call send a message along that route? Hopefully my thoughts are on the right track. Thanks again for the comments!

kellylougheed commented:
Here’s the latest version of my code! Disclaimer: no idea if it works yet. It was mostly written on an airplane whose wifi I refused to pay for.

What I did:
- For the AJAX requests in draw-forfeit.js, I added another AJAX call on the success function that calls a route in the controller: draw, reject_draw, or forfeit.
- I used the js-routes gem that allowed me to put Ruby paths into Javascript: https://github.com/railsware/js-routes (and another helpful article: https://www.sitepoint.com/rails-url-helpers-javascript-jsroutes/)
- In the controller, these routes (draw, reject_draw, forfeit) detect the latest action and the current user’s color and trigger Pusher to display the appropriate message on the appropriate user’s screen. They also trigger Pusher to show certain buttons on the screen (e.g. “forfeit” or “accept draw”/“reject draw”).
- I added event bindings for the messages & buttons in applications.js.coffee.

What I still need to do:
- Find a way to make the line “require ‘pusher’” in pusher.rb NOT crash “bundle exec rspec” and the Codeship build.
- Figure out if this all works and how to test it.

Questions I have:
- I classified the routes for draw, reject_draw, and forfeit as “put” because they seemed more involved with sending data rather than getting it. But I wasn’t sure if it should be “get” or something else. Thoughts?
- Am I on the right path here?

FsDevNinja commented on app/assets/javascripts/application.js:
```diff
@@ -16,3 +16,7 @@
 //= require_tree .
+//= require js-routes
+
+window.pusher = new Pusher(window.pusher_key);
```
I was getting an error in console about the pusher_key not being set right. I switched out `window.pusher_key` to the actual key string and the error went away. not sure where `window.pusher_key` is coming from.

kellylougheed replied:
Hmm, not sure if I got that error. I was storing the key in the environment variables (.env on my local computer) so maybe that's why. But I'm actually not sure that the key needs to be secret (it's in the default script you're supposed to include -- and there's another explicitly secret key) so I'll just hardcore it here as well.

FsDevNinja commented on app/assets/javascripts/application.js.coffee:
```diff
@@ -0,0 +1,23 @@
+  window.broadcast_channel.bind 'draw_forfeit', (event) ->
+    $('#message').html(event.message)
+
+// Show default buttons
```
`//` is not a comment in coffee i dont think. Did get an error for this. Changed it to # and error went away.

FsDevNinja commented on app/assets/javascripts/draw-forfeit.js:
```diff
@@ -0,0 +1,138 @@
+          alert("A draw has been requested!");
+          $.ajax({
+            type: 'PUT',
+            url: Routes.game_draw_path(game_id)
```
I"m not sure how the `Routes.game_draw_path(game_id)` works in js. but I did get an error for game_id. if i hardcoded the game id in there, might be easier just to throw the id in a data-attr and grab it from there like you are on the updateURL.

kellylougheed replied:
Blargh. That was some syntax I found in one tutorial, but it looks like in the actual documentation they do want exact numbers... I'll take your suggestion and put the id in a data-attr!

FsDevNinja commented on config/initializers/pusher.rb:
```diff
@@ -0,0 +1,7 @@
+require 'pusher'
```
you are requiring pusher but not providing it.  You need to add `gem "pusher"` to your Gemfile.

lasley commented on app/views/layouts/application.html.erb:
```diff
@@ -19,6 +19,18 @@
+<script>
+  // Initialize Firebase
+  var config = {
+    apiKey: "AIzaSyDcZeyaMD_zB8BKh8P1ciNhNwrPaFW9lqw",
```
This API key is compromised. Would be a good idea to revoke it

lasley replied:
Oh nvm this was a Firebase key - I thought it was pusher. These are meant to be public, so this was fine.

kellylougheed commented:
Things are looking good! When the user draws, the AJAX call updates the server, and a second AJAX call in the success function hits the controller to trigger Pusher, which displays a message and changes the buttons. More extensive multi-player testing is needed to test the whole process from both players' perspectives, but I'm pretty optimistic.

I got rid of JS routes because it needed ids hardcoded anyway, and was also inexplicably attaching .json to everything.

@FsDevNinja I did not run into the error Pusher::ConfigurationError (missing key `key' in the client configuration) on my computer. But the environment variables will definitely be something to keep an eye on in pushing to Heroku, etc.

PamBWillenz commented:
Wow, @kellylougheed Kelly - this is some intricate JS and great job with pusher. I am in completely unfamiliar territory with JS but am following your steps. Great work - you troubleshooted this until it worked - HUGE SUCCESS!!!!

My one question (just so I know) - is `url: id+"/reject_draw"` pulling from the player's id or the game id?

And on all the JS - how did you know which variables to create? From the methods you created in the game#controller?

kellylougheed commented:
Thanks, @PamBWillenz!! I'm so psyched that it's working. "id" is a JavaScript variable from the data-attr for the game id (as Matt suggested doing). For the JS, I grabbed the update url and game id so that I would be apple to make the AJAX put requests, and I grabbed the player color so that I knew whether it was a black or white draw/forfeit. Thanks for looking at the code -- glad you're following!

kellylougheed commented:
Latest updates:
- Made private channels so that different users would see different button options.
- Added a "pusher" control to allow user authentication into the different channels (the creation of a new controller made coveralls decrease, but I am literally just using it from one route: pusher/auth).
- Added some CSS to make the messages look nice.
- Added a few lines of code on show.html.erb that would let you know if this game had been forfeited or come to a draw, and also tell you what color player you were.
- Changed the event binding in games.js.coffee to be mostly CSS-oriented (displaying and not displaying the appropriate buttons).

Question: What kind of tests are necessary here? The code is very JS-heavy.

Let me know what you guys think, and thanks for taking a look!
